File: acl.go
package main

import "strings"

type ACL interface {
	CanPublish(username, topic string) bool
	CanSubscribe(username, topic string) bool
}

type AllowAllACL struct{}

func (AllowAllACL) CanPublish(username, topic string) bool  { return true }
func (AllowAllACL) CanSubscribe(username, topic string) bool { return true }

type MapACL struct {
	publishAllowed   map[string][]string
	subscribeAllowed map[string][]string
}

func NewMapACL() *MapACL {
	return &MapACL{
		publishAllowed:   make(map[string][]string),
		subscribeAllowed: make(map[string][]string),
	}
}

func (a *MapACL) AllowPublish(username, topicFilter string) {
	a.publishAllowed[username] = append(a.publishAllowed[username], topicFilter)
}

func (a *MapACL) AllowSubscribe(username, topicFilter string) {
	a.subscribeAllowed[username] = append(a.subscribeAllowed[username], topicFilter)
}

func (a *MapACL) CanPublish(username, topic string) bool {
	for _, filter := range a.publishAllowed[username] {
		if matchTopics(strings.Split(topic, "/"), strings.Split(filter, "/")) {
			return true
		}
	}
	return false
}

func (a *MapACL) CanSubscribe(username, topic string) bool {
	for _, filter := range a.subscribeAllowed[username] {
		if matchTopics(strings.Split(topic, "/"), strings.Split(filter, "/")) {
			return true
		}
	}
	return false
}

File: lib/lqtt/acl.ex
defmodule Lqtt.ACL do
  def allow_all do
    fn :publish, _username, _topic -> true
       :subscribe, _username, _topic -> true
    end
  end

  def map_acl(opts \\ []) do
    publish_allowed = Keyword.get(opts, :publish_allowed, %{})
    subscribe_allowed = Keyword.get(opts, :subscribe_allowed, %{})

    fn
      :publish, username, topic ->
        filters = Map.get(publish_allowed, username, [])
        Enum.any?(filters, &match_topics(topic, &1))

      :subscribe, username, topic ->
        filters = Map.get(subscribe_allowed, username, [])
        Enum.any?(filters, &match_topics(topic, &1))
    end
  end

  defp match_topics(topic, filter) do
    topic_parts = String.split(topic, "/")
    filter_parts = String.split(filter, "/")
    do_match_levels(topic_parts, filter_parts, 0)
  end

  defp do_match_levels(_topic_parts, filter_parts, fi) when fi >= length(filter_parts) do
    false
  end

  defp do_match_levels(topic_parts, filter_parts, fi) do
    filter = Enum.at(filter_parts, fi)

    cond do
      filter == "#" ->
        true

      fi >= length(topic_parts) ->
        false

      filter == "+" ->
        do_match_levels(topic_parts, filter_parts, fi + 1)

      Enum.at(topic_parts, fi) == filter ->
        do_match_levels(topic_parts, filter_parts, fi + 1)

      true ->
        false
    end
  end
end
